feat(ui): UI :detach command

justinmk opened the pull request:
(Part of https://github.com/neovim/neovim/pull/30319)

## Problem:

Cannot detach the current UI.

## Solution:

- Introduce `:detach`.
- Introduce `Channel.detach`.

close #23441
close #23093

## TODO

- Windows support is opt-in via `__NVIM_DETACH=1`, but currently broken.

justinmk committed:
feat(ui): UI :detach command

Problem:
Cannot detach the current UI.

Solution:
- Introduce `:detach`.
- Introduce `Channel.detach`.

Co-authored-by: bfredl <bjorn.linse@gmail.com>

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -3554,26 +3643,7 @@ describe('TUI client', function()
-    ]],
-    }
-    -- assert that client has exited
-    screen_client:expect {
```
@zeertzjq this expect() is the same as the one above. Is that intentional?

zeertzjq replied:
It's not the same. The session is different.

justinmk replied:
Ok restored, but see https://github.com/neovim/neovim/pull/32385#discussion_r1948274411

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -3554,26 +3643,7 @@ describe('TUI client', function()
-
-    server_super:close()
-    client_super:close()
+    screen_server:expect({any='Process exited ' .. status})
```
@zeertzjq in https://github.com/neovim/neovim/commit/c26b39a9aa56e834262753e08372240fde9dcdf1 you changed this expect() to be more strict. I relaxed it again because the cirruss freebsd tests fail:

```
FAILED  
test/functional/terminal/tui_spec.lua
 @ 
3666
: 
TUI client exits when server quits with :quit
test/functional/terminal/tui_spec.lua:3646: Row 1 did not match.
Expected:
  |*                                                  |
  |*[Process exited 0]^ {MATCH:%s+}|
  |*                                                  |
  |*                                                  |
  |*                                                  |
  |                                                  |
  |{3:-- TERMINAL --}                                    |
Actual:
  |*Hello, World                                      |
  |*[Process exited 0]{4:^                                }|
  |*{4:~                                                 }|
  |*{4:~                                                 }|
  |*{5:[No Name] [+]                                     }|
  |                                                  |
  |{3:-- TERMINAL --}                                    |
```

Doesn't seem like an essential behavior, although ideally it would be more consistent on all platforms.

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -33,6 +33,92 @@ local assert_log = t.assert_log
+end)
+
 if t.skip(is_os('win')) then
   return
```
Not good that we are skipping all these tests on Windows. But at least the new tests above should run there.

zeertzjq commented on test/functional/core/job_spec.lua:
```diff
@@ -1265,12 +1265,16 @@ describe('jobs', function()
-      {3:-- TERMINAL --}                                    |
-    ]])
+    if is_os('freebsd') then
+      screen:expect { any = '[Process exited 0]' }
```
```suggestion
      screen:expect { any = vim.pesc('[Process exited 0]') }
```

justinmk committed:
test: use --clean

justinmk committed:
fix(ui): Windows :detach is opt-in

Problem:
On Windows, spawning the `nvim --embed` server with `detach=true` breaks
various `tt.setup_child_nvim` tests.

Solution:
Make this behavior opt-in with an env var, temporarily.

justinmk commented:
Can someone with Windows, try this out after defining env var `__NVIM_DETACH=1` ?
